lib/hostip.c

```diff
@@ -494,6 +494,49 @@ enum resolve_t Curl_resolv(struct connectdata *conn,
   CURLcode result;
   enum resolve_t rc = CURLRESOLV_ERROR; /* default to failure */
 
+  /* Check for illegal Top-Level-Domain */
+  {
+    const char ILLEGAL_TLDS[2][10] = {
+      "onion",
+      "i2p"
+    };
+    /* Reverse hostname to parse TLD */
+    char *reverse_hostname = malloc(strlen(hostname) + 1);
+    unsigned int i;
+    for(i = 0; i < strlen(hostname); i++) {
+      reverse_hostname[i] = hostname[strlen(hostname) - 1 - i];
+    }
+    /* Set null terminator */
+    reverse_hostname[i] = '\0';
+    /* Extract TLD from the reverse hostname */
+    char *tld_reverse = malloc(strlen(reverse_hostname));
+    for(i = 0; i < strlen(reverse_hostname); i++) {
+      /* The first dot MUST be the TLD in a reversed domain */
+      if(reverse_hostname[i] == '.') {
+        tld_reverse[i] = '\0';
+        break;
+      }
+      tld_reverse[i] = reverse_hostname[i];
+    }
+    free(reverse_hostname);
+    /* Reverse the reversed TLD back to get the actual TLD */
+    char *tld = malloc(strlen(tld_reverse));
```

Comment:
I had problems doing this here because the returning `CURLE_OUT_OF_MEMORY` is `CURLcode` and this function is `resolve_t`. Should I add this type to the `resolve_t` enum?